allennlp/training/trainer.py

```diff
@@ -191,15 +213,83 @@ def on_batch(
         batch_number: int,
         is_training: bool,
         is_primary: bool = True,
+        batch_grad_norm: Optional[float] = None,
         **kwargs,
     ) -> None:
         # In the distributed case we need to call this from every worker, since every
         # worker reports its own memory usage.
         cpu_memory_usage = common_util.peak_cpu_memory()
         gpu_memory_usage = common_util.peak_gpu_memory()
-        # But we only want to log from the primary process.
+
+        if not is_primary:
+            return None
+        assert self._tensorboard is not None
+
+        if self._tensorboard.should_log_histograms_next_batch():
+            # Get the magnitude of parameter updates for logging.  We need to do some
+            # computation before and after the optimizer step, and it's expensive because of
+            # GPU/CPU copies (necessary for large models, and for shipping to tensorboard), so
+            # we don't do this every batch, only when it's requested.
+            self._param_updates = {
+                name: param.detach().cpu().clone()
+                for name, param in trainer.model.named_parameters()
+            }
+        elif self._tensorboard.should_log_histograms_this_batch():
+            assert self._param_updates is not None
+            for name, param in trainer.model.named_parameters():
+                self._param_updates[name].sub_(param.detach().cpu())
```

Comment:
These conditions could both be true at the same time, if we log histograms every batch. So `this_batch` needs to be processed first, and the `if` statements need to be independent.

Reply:
Good catch. Fixed